### DISCUSSION TITLE
Port 25 or Port 587

### DISCUSSION THREAD

**Discussion by impirecom** [postalserver/postal#2563] [Help with using Postal] [answered]
Hello,

Can Postal support multiple ports? 

Also, will port 587 provide more security by default? Per this article it does:

https://www.cloudflare.com/learning/email-security/smtp-port-25-587/

**Comment by willpower232** [comment]
Postal can only bind to a single port however there is nothing stopping you routing multiple ports within the server with iptables. If you search the issues/discussions for 2525 then you should be able to find an example of it.

The only thing Postal doesn't support is implicit TLS, all connections to Postal must upgrade to TLS via STARTTLS after initially connecting insecurely.

I wouldn't say that just using a different port number provides any additional security as in Postal case it is the same service and code receiving the network traffic. It mostly depends on any network blocks you need to work around.

**Reply by impirecom** [reply]
Thank you very much, but I am confused by this "The only thing Postal doesn't support is implicit TLS, all connections to Postal must upgrade to TLS via STARTTLS after initially connecting insecurely."

I see Postal have the option to enable TLS, this mean it support STARTTLS? If so, what are the steps to get it to work. Is there a step-by-step guide to get this to work, including how to install the certificate?

**Reply by willpower232** [reply]
If you can provide a certificate to Postal via the docker-compose file then you can enable TLS in Postal. There are many discussions about setting it up, have a look at the end of #98 

Some ports (465 I think) are for SMTPS which has TLS by default so if you use that port then you will probably have problems as Postal doesn't talk in TLS by default.

**Reply by impirecom** [reply]
Having problem with getting SSL to work. eMail client cannot connect. Can't seems to figure out what's wrong:

### MY CONFIG:

_smtp_server:
  port: 587
  tls_enabled: true
  ssl_mode: STARTTLS
  tls_certificate_path: /opt/postal/config/smtp.cert
  tls_private_key_path: /opt/postal/config/smtp.key_

**I did as instructed from https://github.com/postalserver/postal/issues/98:**

root@host1:~# sudo certbot certonly --standalone -d postal.yourdomain.com --key-type rsa cp /etc/letsencrypt/live/postal.yourdomain.com/fullchain.pem smtp.cert cp /etc/letsencrypt/live/postal.yourdomain.com/privkey.pem smtp.key root@host1:/opt/postal/config# ls -lht

verify it openssl s_client -starttls smtp -crlf -connect postal.yourdomain.com:587

**Received result:**

CONNECTED(00000003)
write:errno=104

no peer certificate available

No client certificate CA names sent

SSL handshake has read 136 bytes and written 351 bytes
Verification: OK

New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)

**Reply by willpower232** [reply]
Unfortunately the config directory is in a different location inside the docker container as per the docker-compose files

https://github.com/postalserver/install/blob/ce3ddfa66e18a7509ebec9c7aebf84fcd70675df/templates/docker-compose.yml#L8

this means your paths in the config file should start with `/config` and not `/opt/postal/config`

You've also specified ssl_mode which doesn't seem to be a setting in this part of the configuration.

Finally after fixing the path and removing the ssl_mode, don't forget to restart Postal.

**Reply by impirecom** [reply]
I am grateful for your help.

Tried with what you suggested, but client still can't connect. Openssl test command also came back with same result as before. Email clients still cannot connect. SMTP tests failed.

FYI: The two files smtp.key and smtp.key are located in /opt/postal/config/

**Current config below I wasn't sure if /config/ or just config/. Does it matter?**

smtp_server:
  port: 587
  tls_enabled: true
  tls_certificate_path: config/smtp.cert
  tls_private_key_path: config/smtp.key

**root@postal:~# openssl s_client -starttls smtp -crlf -connect postal.yourdomain.com:587**

CONNECTED(00000003)
write:errno=104
no peer certificate available
No client certificate CA names sent
SSL handshake has read 136 bytes and written 351 bytes
Verification: OK
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)

**Reply by willpower232** [reply]
Judging from the install repo, it should be /config.

If that doesn't fix it then it might be worth trying port 25 which you may have better experience with long term anyway.

**Reply by impirecom** [reply]
I have tried /config as the parameter, still same issue. I also tried port 25. But strangely, first time see different error. Ran command again, shown same error as before:

  port: 25
  tls_enabled: true
  tls_certificate_path: /config/smtp.cert
  tls_private_key_path: /config/smtp.key

Something is off. Seems like it doesn't even see the certificate and key:

root@vm1:~# openssl s_client -starttls smtp -crlf -connect postal.example.com:25

_4007C6D6C67F0000:error:8000006F:system library:BIO_connect:Connection refused:../crypto/bio/bio_sock2.c:125:calling connect()
4007C6D6C67F0000:error:10000067:BIO routines:BIO_connect:connect error:../crypto/bio/bio_sock2.c:127:
connect:errno=111_

root@vm1:~# openssl s_client -starttls smtp -crlf -connect postal.example.com:25

_CONNECTED(00000003)
write:errno=104
no peer certificate available
No client certificate CA names sent
SSL handshake has read 136 bytes and written 351 bytes
Verification: OK
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)_

Anyway, I do need this to work on port 587. some apps that I uses support port 587 and block port 25 (Hexospark, Instantly and Mystrika as examples). I have no choice but to use port 587. When I asked them to open port 25, they sent me the below article and told to use better practice:

https://www.cloudflare.com/learning/email-security/smtp-port-25-587/

I am back to square one and losing hairs (what's left of i). At this point, I would like to pay someone to implement this. Couldn't understand why Postal made it so hard for users to use port 587 with SSL.

**Reply by willpower232** [reply, marked as answer]
if you've copied the certificate files to the config directory, you could try `postal bash smtp` and `ls -hal /config` to confirm the permissions on the files. It may be that they're restricted to root or something that prevents Postal from reading it.

**Reply by impirecom** [reply]
That worked. It was the permission issue. chmod command solved it. Thank you much.

**Reply by impirecom** [reply]
I am not sure if this is related. The day after I added SSL and enabled port 587, incoming mails showing as zero. Outgoing still works but strangely server show no incoming messages. 
<img width="608" alt="Screenshot 2023-08-29 at 8 15 27 AM" src="https://github.com/postalserver/postal/assets/79177007/16099e8b-a93d-4640-b7f6-d62604483ccd">

**Reply by willpower232** [reply]
if you no longer have port 25 routed to Postal, it is highly likely that incoming messages are not making it to Postal at all

**Reply by impirecom** [reply]
Thank you Will. You are right, mail relay to the smtp connections are not working. How do I route port 25 to postal?

Also, do you have an instruction on creating a cron job to auto-reboot the servers to avoid Let's Encrypt SSL expired issue?

I was wondering how I can install cloudflare SSL, which can last up to 15 years. This should eliminate the need to reboot the servers every 90 days.

**Reply by willpower232** [reply]
> How do I route port 25 to postal?
You can use iptables to route multiple ports to the same port. Its somewhere in the issues/discussions here

> auto-reboot the servers
This depends how you are obtaining the certificate. If you're using certbot directly, then I think it has some kind of after-renewal hook thing which you can use. If you're getting them from another service, you'd have to see what callbacks they had available.

> cloudflare SSL
This would only work if they give you both parts/files for the certificate I believe. I'm not sure if they do that or keep the certificate to themselves.

**Comment by mostafijar** [comment]
https://www.youtube.com/watch?v=F5x5NxUU5Pc
postal with TLS
